content/blog_posts/articles/roadmap-to-adonisjs-7.md

```diff
@@ -0,0 +1,389 @@
+
+In the following example, a transformer is defined for the `Poll` model. The `toObject` method returns the serialized representation of the model, ready to be sent in an HTTP response.
+
+```ts
+import { BaseTransformer } from '@adonisjs/core/transformers';  
+import PollOptionTransformer from '#transformers/poll_option';  
+import UserTransformer from '#transformers/user';  
+import type Poll from '#models/poll';
+
+export default class PollTransformer extends BaseTransformer<Poll> {
+  toObject() {
+    return {
+      /**
+       * Pick properties from an object (a model instance in this case)
+       */
+      ...this.pick(this.resource, [
+        'id',
+        'title',
+        'pollColor',
+        'slug',
+        'expired',
+      ]),
+      /**
+       * Additional properties to serialize
+       */
+      closesAt: this.resource.closesAt.toRelative({ style: 'short' }),
+      votesCount: this.countOf('votes'),
+      /**
+       * Use other transformers
+       */
+      author: UserTransformer.item(this.whenLoaded(this.resource.author)),
+      options: PollOptionTransformer.collection(this.whenLoaded(this.resource.options)),
+    };
+  }
+}
+```
+
+### Infering Type
+
+If you're building an Inertia app or an API used in the same codebase as your frontend, you will be able to generate types directly from your transformers. This will allow your frontend to consume backend data with full type safety.
+
+Generated types are stored in `.adonisjs/frontend/data.d.ts`:
+
+```ts
+import { InferData } from '@adonisjs/core/transformers';
+
+export type PollOptionData = InferData<import('../../app/transformers/poll_option.ts').default>;
+export type PollData = InferData<import('../../app/transformers/poll.ts').default>;
+export type UserData = InferData<import('../../app/transformers/user.ts').default>;
+```
+
+### Notes on Transformers
+- Transformers can be used with **any data type**, not just Lucid models. 
+- The `toObject` method **must return JSON-serializable values** (no classes, maps, sets, or symbols).
+- `toObject` can be **async**, allowing you to embed things like permissions via bouncer policies.
+- Since transformers are resolved through the **IoC container**, you will be able to inject services using the `@inject` decorator.
+
+## Type-Safe URL Generator
+
+AdonisJS 7 will introduce a **type-safe URL builder** for generating route URLs, both on the backend and the frontend.
+
+You will be able to generate a client for your frontend application automatically. This client will live inside the `.adonisjs/frontend/url_builder.ts` file and offer the same functional API as on the server side, giving you full auto-completion and type safety across your entire stack.
+
+Here's what the developer experience will look like:
+
+```ts
+import { route } from '@adonisjs/core/http/url_builder';
```

Comment:
Should we import `urlFor` here instead. Next, we can ask for feedback on naming and give users choice between `route`, `url`, and `urlFor`. I can write a small explanation for all of these naming options